lib/model/content.dart

```diff
@@ -658,27 +672,54 @@ class _ZulipContentParser {
       return UnimplementedBlockContentNode(htmlNode: divElement);
     }
 
-    final buffer = StringBuffer();
+    final spans = <CodeBlockSpanNode>[];
     for (int i = 0; i < mainElement.nodes.length; i++) {
+      final CodeBlockSpanNode span;
+
       final child = mainElement.nodes[i];
-      if (child is dom.Text) {
-        String text = child.text;
-        if (i == mainElement.nodes.length - 1) {
-          // The HTML tends to have a final newline here.  If included in the
-          // [Text] widget, that would make a trailing blank line.  So cut it out.
-          text = text.replaceFirst(RegExp(r'\n$'), '');
-        }
-        buffer.write(text);
-      } else if (child is dom.Element && child.localName == 'span') {
-        // TODO(#191) parse the code-highlighting spans, to style them
-        buffer.write(child.text);
-      } else {
-        return UnimplementedBlockContentNode(htmlNode: divElement);
+      switch (child) {
+        case dom.Text(:var text):
+          if (i == mainElement.nodes.length - 1) {
+            // The HTML tends to have a final newline here.  If included in the
+            // [Text] widget, that would make a trailing blank line.  So cut it out.
+            text = text.replaceFirst(RegExp(r'\n$'), '');
+          }
+          span = CodeBlockSpanNode(text: text, type: CodeBlockSpanType.text);
+
+        case dom.Element(:final text, :final classes, :final localName)
+          when localName == 'span'
+            // We assume that server will always provide us with a span element
+            // that has at most one class associated with the span element.
+            && classes.length <= 1:
+          final CodeBlockSpanType type;
+          if (classes.isEmpty) {
+            type = CodeBlockSpanType.text;
+          } else {
+            type = codeBlockSpanTypeFromClassName(classes.first);
+          }
+          switch (type) {
+            case CodeBlockSpanType.unknown:
+              return UnimplementedBlockContentNode(htmlNode: divElement);
+            case CodeBlockSpanType.highlightedLines:
+              // TODO: Implement nesting in CodeBlockSpanNode to support hierarchically
+              //       inherited styles for `span.hll` nodes.
+              return UnimplementedBlockContentNode(htmlNode: divElement);
+            default:
+              span = CodeBlockSpanNode(text: text, type: type);
+          }
+
+        default:
+          return UnimplementedBlockContentNode(htmlNode: divElement);
+      }
+      if (span.text.isNotEmpty) {
+        // We don't want to emit unnecessary CodeBlockSpanNode with empty text
+        // which results in a TextSpan with empty text which results in
+        // nothing displayed in the end anyway.
```

Comment:
This happens on every syntax highlighted code block, probably because of the newline being replaced with empty string in `dom.Text()` case.

Reply:
In previous revisions this if check was inside the `dom.Text()` case but I moved it out when adding the `final CodeBlockSpanNode span;` because dart doesn't like that we don't assign it in the else case.
Also this way if the `dom.Element()` case results in a node with empty string then that node will get ignored too.